doc/design/zarr-support.md

```diff
@@ -0,0 +1,100 @@
+
+Given these considerations here are questions for implementation
+1. Is there a way to upload a folder to a given prefix using an API key without having to create 100k signed URLs?
+1. Should the tree structure be stored somewhere so that diffs can be ascertained?
+   * Assuming 64 bytes of checksum, 64 bytes for path (optimistic), and 1million files (pessimistic), that's 128MB of checksum info in the DB. Plausible for upload, but probably not long term.
+1. Given that each zarr file may contain 100k+ files, how will dandi-cli handle alterations?
+
+# Requirements
+
+1. zarr files are stored in a "directory" in S3.
+1. Each zarr file corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr file. The API verifies this checksum _immediately_ after upload; it's not good enough to download the entire zarr file to calculate it after upload.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+
+# Implementation
+
+## Hash function for checksums
+The hash function used to generate the checksum for the zarr file must have these characteristics:
+
+1. It must be based on the dandi-etag function of the subfile contents, and the path of the subfile in the zarr file.
+1. A well-defined ordering of the subfiles in the zarr file must exist.
+The checksum must be computed on the subfiles in this order.
+1. It must be applicable on one subfile at a time.
+1. It must be able save its state between subfiles.
+
+**TODO** what is this function?
+
+A simple scheme that I think would work:
+
+* initial value is `sha256("{etag}:{path}")` for the first subfile
+* the next value is `sha256("{prev_sha256}:{etag}:{path}")`, ad infinitum
+
+We could also do something more complicated with merkle trees or something, but it would be much more obnoxious to compute.
+
+## Upload flow
+
+### Before upload (technically optional)
+1. CLI calculates the checksum of the zarr file.
+1. CLI queries the API (**TODO** URL?) if the checksum has already been uploaded.
+1. If so, it proceeds with the already uploaded zarr file and skips upload entirely.
+
+### Upload
+1. CLI queries the API (**TODO** URL?) with the checksum of the zarr file to initiate an upload. The API creates an upload UUID, records the checksum, and initializes a "running checksum" to `null`.
+1. CLI requests a batch of presigned URLs (**TODO** URL?).
+The files must be ordered in the same way used to calculate the checksum.
+Any number of files may be requested in a batch.
+The request body looks like this:
+```
+[
+   {
+      "path": "foo/bar.txt",
+      "dandi-etag": "abc ... 123"
+   },
+   {
+      "path": "foo/baz.txt",
+      "dandi-etag": "def ... 456"
+   },
+]
+```
+3. API responds with a corresponding list of presigned upload URLs (**TODO** where to upload?) in the S3 bucket.
+The size limit for each upload is 5GB.
```

Comment:
`dandi-etag` is the S3 etag, which for this case is just the MD5 of the file, so it already is, basically. Do you mean that we should use the etag to generate the presigned URL so that only a file with that etag can be uploaded?

Reply:
> so that only a file with that etag can be uploaded
yes -  we couldn't do it for multipart, but should work for single part.